knowledge-base/grid-custom-filter-menu.md

```diff
@@ -0,0 +1,201 @@
+
+This knowledge base article also answers the following questions:
+- How to integrate AutoComplete with `StringOperators` in a TelerikGrid `FilterMenuTemplate`?
+- How to refresh a custom `FilterMenuTemplate` in TelerikGrid?
+- How to refresh a `FilterMenuTemplate` context after programmatic changes?
+
+## Solution
+
+To achieve the desired behavior, encapsulate the content of the `FilterMenuTemplate` in a custom component and refresh this component upon selection change in the dropdown list. Below is an example demonstrating this approach:
+
+<div class="skip-repl"></div>
+````Home.razor
+@using Telerik.DataSource
+<TelerikGrid Data="@Countries"
+             FilterMode="GridFilterMode.FilterMenu"
+             PageSize="25">
+    <GridColumns>
+        <GridColumn Field="@(nameof(Country.CountryName))">
+            <FilterMenuTemplate>
+                <CustomFilterMenu @bind-FilterOperator="@FilterOperator"
+                                  @bind-SelectedCountry="@SelectedCountry"
+                                  Countries="@Countries"
+                                  FilterOperators="@FilterOperators" />
+            </FilterMenuTemplate>
+            <FilterMenuButtonsTemplate Context="filterContext">
+                <TelerikButton OnClick="@(async _ => await FilterAsync(filterContext))" Class="k-button-solid-primary">Filter</TelerikButton>
+                <TelerikButton OnClick="@(async _ => await ClearFilterAsync(filterContext))">Clear</TelerikButton>
+            </FilterMenuButtonsTemplate>
+        </GridColumn>
+    </GridColumns>
+</TelerikGrid>
+
+@code {
+    private IEnumerable<StringFilterOperator> FilterOperators { get; set; } = Enum.GetValues(typeof(StringFilterOperator)).Cast<StringFilterOperator>().ToList();
+    private StringFilterOperator FilterOperator { get; set; } = StringFilterOperator.StartsWith;
+
+    private string SelectedCountry { get; set; } = null!;
+
+    private async Task FilterAsync(FilterMenuTemplateContext filterContext)
+    {
+        FilterHandler(filterContext.FilterDescriptor);
+        await filterContext.FilterAsync();
+    }
+
+    private async Task ClearFilterAsync(FilterMenuTemplateContext filterContext)
+    {
+        SelectedCountry = string.Empty;
+
+        FilterHandler(filterContext.FilterDescriptor);
+        filterContext.FilterDescriptor.FilterDescriptors.Clear();
+        await filterContext.ClearFilterAsync();
+    }
+    private void FilterHandler(CompositeFilterDescriptor filterDescriptor)
+    {
+        var model = string.Empty;
+        object? value = null;
+
+        value = SelectedCountry;
+        model = nameof(Country.CountryName);
+
+        filterDescriptor.FilterDescriptors.Clear();
+        filterDescriptor.FilterDescriptors.Add(new FilterDescriptor(model, Telerik.DataSource.FilterOperator.IsEqualTo, value));
+    }
+
+    private List<Country> Countries { get; set; } = new List<Country>()
+{
+    new Country { Id = 1, CountryName = "Albania" },
```

Comment:
Consider random data that is generated in a loop. Will make the code a lot shorter. In this case, it makes sense to have a different string in each data item and this is possible too. If you ensure that every letter and digit is represented, you will be fine. For example - https://docs.telerik.com/blazor-ui/components/grid/filter/searchbox